inception-ui-external-search/src/main/java/de/tudarmstadt/ukp/inception/app/ui/externalsearch/SearchPage.java

```diff
@@ -334,6 +336,15 @@ public ResultRowView(String id, long rowNumber, IModel<ExternalSearchResult> mod
             add(new LambdaAjaxLink("importLink", _target -> actionImportDocument(_target, result))
                     .add(visibleWhen(() -> 
                         !documentService.existsSourceDocument(project, documentTitle))));
+            add(new LambdaAjaxLink("openLink", _target -> {
+                PageParameters pageParameters = new PageParameters()
+                    .add(WebAnnoConst.PAGE_PARAM_PROJECT_ID, project.getId())
+                    .add(WebAnnoConst.PAGE_PARAM_DOCUMENT_ID,
+                        documentService.getSourceDocument(project, documentTitle).getId())
+                    .add(WebAnnoConst.PAGE_PARAM_FOCUS, 1);
+                setResponsePage(AnnotationPage.class, pageParameters);
+            }).add(
+                visibleWhen(() -> documentService.existsSourceDocument(project, documentTitle))));
```

Comment:
We call `documentService.existsSourceDocument` three times in this method. It is probably a cheap operation for most plausible `DocumentService` implementations, but there's no harm in just doing it once and saving the result.

Reply:
When does the `visibleWhen` check get repeated? Can you confirm this updates appropriately when the user imports a document?

Reply:
Followed the suggestion and call `documentService.existsSourceDocument` once.

Reply:
The visibility is updated every time the component is re-rendered. As long as either the entire page is reloaded or if the component is added directly or indirectly to an `AjaxRequestTarget`, it is re-rendered. If a component needs to be become visible during an AJAX request, it is necessary that `setOutputMarkupPlaceholderTag(true)` was called on it so that the AJAX request finds the element in the browser-side HTML page and can re-render it.